st2auth/st2auth/handlers.py

```diff
@@ -161,17 +163,53 @@ def handle_auth(self, request, headers=None, remote_addr=None, remote_user=None,
         username, password = split
 
         result = self._auth_backend.authenticate(username=username, password=password)
+
         if result is True:
             ttl = getattr(request, 'ttl', None)
             username = self._get_username_for_request(username, request)
             try:
-                token = self._create_token_for_user(
-                    username=username, ttl=ttl)
-                return token
+                token = self._create_token_for_user(username=username, ttl=ttl)
             except TTLTooLargeException as e:
                 abort_request(status_code=http_client.BAD_REQUEST,
                               message=e.message)
                 return
 
+            # If remote group sync is enabled, sync the remote groups with local StackStorm roles
+            if cfg.CONF.rbac.sync_remote_groups:
+                LOG.debug('Retrieving auth backend groups for user "%s"' % (username),
+                          extra=extra)
+                try:
+                    user_groups = self._auth_backend.get_user_groups(username=username)
+                except NotImplementedError:
+                    LOG.debug('Configured auth backend doesn\'t expose user group membership '
+                              'information, skipping sync...')
+                    return token
+
+                if not user_groups:
+                    # No groups, return early
+                    return token
+
+                extra['username'] = username
+                extra['user_groups'] = user_groups
+
+                LOG.debug('Found "%s" groups for user "%s"' % (len(user_groups), username),
+                          extra=extra)
+
+                user_db = UserDB(name=username)
+                syncer = RBACRemoteGroupToRoleSyncer()
+
+                try:
+                    syncer.sync(user_db=user_db, groups=user_groups)
+                except Exception as e:
+                    # Note: Failed sync is not fatal
+                    LOG.exception('Failed to synchronize remote groups for user "%s"' % (username),
+                                  extra=extra)
+                else:
+                    LOG.debug('Successfuly synchronized groups for user "%s"' % (username),
+                              extra=extra)
+
+                return token
+            return token
```

Comment:
I actually personally usually prefer to return early, but I just followed the existing style so I went with this approach.

I'm fine either way aka fine with changing it.